cpp/src/sort/segmented_sort.cu

```diff
@@ -35,24 +34,49 @@ namespace {
  */
 enum class sort_method { STABLE, UNSTABLE };
 
-// returns segment indices for each element for all segments.
-// first segment begin index = 0, last segment end index = num_rows.
+/**
+ * @brief Builds indices to identify segments to sort
+ *
+ * The segments are added to the input table-view keys so they
+ * are lexicographically sorted within the segmented groups.
+ *
+ * ```
+ * Example 1:
+ * num_rows = 10
+ * offsets = {0, 3, 7, 10}
+ * segment-indices -> { 3,3,3, 7,7,7,7, 10,10,10 }
+ * ```
+ *
+ * ```
+ * Example 2: (offsets do not cover all indices)
+ * num_rows = 10
+ * offsets = {3, 7}
+ * segment-indices -> { 0,1,2, 7,7,7,7, 8,9,10 }
+ * ```
+ *
+ * @param num_rows Total number of rows in the input keys to sort
+ * @param offsets The offsets identifying the segments
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ */
 rmm::device_uvector<size_type> get_segment_indices(size_type num_rows,
                                                    column_view const& offsets,
                                                    rmm::cuda_stream_view stream)
 {
   rmm::device_uvector<size_type> segment_ids(num_rows, stream);
 
-  auto offset_begin = offsets.begin<size_type>();  // assumes already offset column contains offset.
-  auto offsets_minus_one = thrust::make_transform_iterator(
-    offset_begin, [offset_begin] __device__(auto i) { return i - 1; });
+  auto offset_begin  = offsets.begin<size_type>();
+  auto offset_end    = offsets.end<size_type>();
   auto counting_iter = thrust::make_counting_iterator<size_type>(0);
-  thrust::lower_bound(rmm::exec_policy(stream),
-                      offsets_minus_one,
-                      offsets_minus_one + offsets.size(),
-                      counting_iter,
-                      counting_iter + segment_ids.size(),
-                      segment_ids.begin());
+  thrust::transform(rmm::exec_policy(stream),
```

Comment:
This function call be removed completely (maybe?) and replaced by `cudf::lists::detail::generate_labels`.

Reply:
I meant this function `get_segment_indices` can be removed and we just call `cudf::lists::detail::generate_labels`. This function output is just appended as an additional column to the input list so we can sort the elements within each lists.

Reply:
The output of `cudf::lists::detail::generate_labels` is not exactly the same as this, but that doesn't matter-we just need to have the same (unique) value for elements of each list.

Reply:
Except `generate_labels` does not produce the same results as per this PR about indices not included in the offsets.
Maybe you can point out where those values are handled? I'm not seeing it.

Reply:
Oh sorry, maybe we can't use it out of the box as we may have missing offsets at both ends. Anyway, if we can reuse anything from these function then it's good.

Reply:
I'm thinking maybe `generate_labels` is a subset of this function which is more efficient (single kernel). So may it can be replaced with this one at some point. I'm not sure how important the actual values are to the callers of `generate_labels`.

Reply:
When implementing `generate_labels` I tried to optimize it by avoiding `upper_bound` and got some (small) speed up. But it can't handle missing offsets thus unfortunately my proposal above (removing this function) can't be adopted right away. It could replace this function completely before this PR though.

Maybe I'll consider modifying `generate_labels` some later time to support missing offsets.